KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+     * @return req min required number of approvals
+     */
+    function _getState(
+        address _retiredAddress
+    ) private returns (address[] memory adminList, uint256 req) {
+        //call getState() function in retiredAddress contract to get the adminList
+        bytes memory payload = abi.encodeWithSignature("getState()");
+        (bool success, bytes memory result) = _retiredAddress.staticcall(
+            payload
+        );
+        emit GetState(success, result);
+        require(success, "call failed");
+
+        (adminList, req) = abi.decode(result, (address[], uint256));
+    }
+
+    /**
+     * @dev Internal function to update the approver details of a retired
+     * _retiredAddress is the address of the retired
+     * _approver is the admin of the retiredAddress
+     */
+    function _updateApprover(
+        address _retiredAddress,
+        address _approver
+    ) private {
+        uint256 index = getRetiredIndex(_retiredAddress);
+        address[] memory approvers = retirees[index].approvers;
+        for (uint256 i = 0; i < approvers.length; i++) {
+            require(
+                approvers[i] != _approver,
+                "Already approved"
+            );
+        }
+        retirees[index].approvers.push(_approver);
+        emit Approved(
+            _retiredAddress,
+            _approver,
+            retirees[index].approvers.length
+        );
+    }
+
+    /**
+     * @dev finalizeRegistration sets the status to Registered,
+     *      After this stage, registrations will be restricted.
+     */
+    function finalizeRegistration()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Initialized)
+    {
+        status = Status.Registered;
+        emit SetStatus(status);
+    }
+
+    /**
+     * @dev finalizeApproval sets the status to Approved,
+     *      After this stage, approvals will be restricted.
+     */
+    function finalizeApproval() public onlyOwner onlyAtStatus(Status.Registered) {
+        require(
+            getTreasuryAmount() < sumOfRetiredBalance(),
+            "treasury amount should be less than the sum of all retired address balances"
+        );
+        _isRetiredsApproved();
+        status = Status.Approved;
+        emit SetStatus(status);
+    }
+
+    /**
+     * @dev verify if quorom reached for the retired approvals
+     */
+    function _isRetiredsApproved() private {
+        for (uint256 i = 0; i < retirees.length; i++) {
+            Retired memory retired = retirees[i];
+            (, uint256 req) = _getState(retired.retired);
+            if (retired.approvers.length >= req) {
+                //if min quorom reached, make sure all approvers are still valid
+                address[] memory approvers = retired.approvers;
+                uint256 minApprovals = 0;
+                for (uint256 j = 0; j < approvers.length; j++) {
+                    _validateAdmin(retirees[i].retired, approvers[j]);
```

Comment:
I think we can optimize this statement. In this function, it also calls _getState() to get admin addresses. Do we need to call _getState() multiple times?

Reply:
Optimised it by removing the `_getState` method from the  _validateAdmin function. So it wont be called twice.